src/ray/object_manager/spilled_object.cc

```diff
@@ -0,0 +1,195 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ray/object_manager/spilled_object.h"
+
+#include <fstream>
+#include <regex>
+
+#include "ray/util/logging.h"
+
+namespace ray {
+namespace {
+const size_t UINT64_size = sizeof(uint64_t);
+}
+
+/* static */ absl::optional<SpilledObject> SpilledObject::CreateSpilledObject(
+    const std::string &object_url, uint64_t chunk_size) {
+  if (chunk_size == 0) {
+    RAY_LOG(WARNING) << "chunk_size can't be 0.";
+    return absl::optional<SpilledObject>();
+  }
+
+  std::string file_path;
+  uint64_t object_offset = 0;
+  uint64_t object_size = 0;
+
+  if (!SpilledObject::ParseObjectURL(object_url, file_path, object_offset, object_size)) {
+    RAY_LOG(WARNING) << "Failed to parse spilled object url: " << object_url;
+    return absl::optional<SpilledObject>();
+  }
+
+  uint64_t data_offset = 0;
+  uint64_t data_size = 0;
+  uint64_t metadata_offset = 0;
+  uint64_t metadata_size = 0;
+  rpc::Address owner_address;
+
+  std::ifstream is(file_path, std::ios::binary);
+  if (!is ||
+      !SpilledObject::ParseObjectHeader(is, object_offset, data_offset, data_size,
+                                        metadata_offset, metadata_size, owner_address)) {
+    RAY_LOG(WARNING) << "Failed to parse object header for spilled object " << object_url;
+    return absl::optional<SpilledObject>();
+  }
+
+  return absl::optional<SpilledObject>(SpilledObject(
+      std::move(file_path), object_size, data_offset, data_size, metadata_offset,
+      metadata_size, std::move(owner_address), chunk_size));
+}
+
+uint64_t SpilledObject::GetDataSize() const { return data_size_; }
+
+uint64_t SpilledObject::GetMetadataSize() const { return metadata_size_; }
+
+const rpc::Address &SpilledObject::GetOwnerAddress() const { return owner_address_; }
+
+uint64_t SpilledObject::GetNumChunks() const {
+  return (data_size_ + metadata_size_ + chunk_size_ - 1) / chunk_size_;
+}
+
+absl::optional<std::string> SpilledObject::GetChunk(uint64_t chunk_index) const {
+  // The spilled file stores metadata before data. But the GetChunk needs to
+  // return data before metadata. We achieve by first read from data section,
+  // then read from metadata section.
+  auto cur_chunk_offset = chunk_index * chunk_size_;
+  auto cur_chunk_size =
+      std::min(chunk_size_, data_size_ + metadata_size_ - cur_chunk_offset);
+
+  std::string result(cur_chunk_size, '\0');
+  size_t result_offset = 0;
+
+  if (cur_chunk_offset < data_size_) {
+    // read from data section.
+    auto offset = cur_chunk_offset;
+    auto size = std::min(data_size_ - cur_chunk_offset, cur_chunk_size);
+    if (!ReadFromDataSection(offset, size, &result[result_offset])) {
+      return absl::optional<std::string>();
+    }
+    result_offset = size;
+  }
+
+  if (cur_chunk_offset + cur_chunk_size > data_size_) {
```

Comment:
Pretty unfortunate we have to do these weird tricks due to the format, but I can't think of a simpler way while preserving the existing IO worker.

If we also used C++ to write local spill files, then I suppose this part could be simplified, but that's a bigger change.